From 2da7fe668973eeeb828ad59ab68189cf410e7460 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 3 Apr 2015 03:58:47 +0200
Subject: [PATCH 0699/1562] server: Allow to open files without any permission
 bits. (try 2)

Changes in v2:
* As suggested by Piotr, fix the problem for both files and directories.
* Pay attention to requested access attributes - this fixes a couple more todo_wine's.

Wine-Staging: server-File_Permissions
---
 dlls/advapi32/tests/security.c | 32 ++++++++++++--------------------
 server/fd.c                    | 21 +++++++++++++++++++++
 2 files changed, 33 insertions(+), 20 deletions(-)

# diff --git a/dlls/advapi32/tests/security.c b/dlls/advapi32/tests/security.c
# index 2840f6bd75a..9569272e0db 100644
# --- a/dlls/advapi32/tests/security.c
# +++ b/dlls/advapi32/tests/security.c
# @@ -3782,17 +3782,13 @@ static void test_CreateDirectoryA(void)
#      error = pGetNamedSecurityInfoA(tmpfile, SE_FILE_OBJECT,
#                                     OWNER_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION,
#                                     (PSID *)&owner, NULL, &pDacl, NULL, &pSD);
# -    todo_wine
#      ok(error == ERROR_SUCCESS, "GetNamedSecurityInfo failed with error %ld\n", error);
# -    if (error == ERROR_SUCCESS)
# -    {
# -        bret = GetAclInformation(pDacl, &acl_size, sizeof(acl_size), AclSizeInformation);
# -        ok(bret, "GetAclInformation failed\n");
# -        todo_wine
# -        ok(acl_size.AceCount == 0, "GetAclInformation returned unexpected entry count (%ld != 0).\n",
# -                                   acl_size.AceCount);
# -        LocalFree(pSD);
# -    }
# +    bret = GetAclInformation(pDacl, &acl_size, sizeof(acl_size), AclSizeInformation);
# +    ok(bret, "GetAclInformation failed\n");
# +    todo_wine
# +    ok(acl_size.AceCount == 0, "GetAclInformation returned unexpected entry count (%ld != 0).\n",
# +                               acl_size.AceCount);
# +    LocalFree(pSD);
#      CloseHandle(hTemp);
 
#      /* Test inheritance of ACLs in NtCreateFile without security descriptor */
# @@ -3861,17 +3857,13 @@ static void test_CreateDirectoryA(void)
#      error = pGetNamedSecurityInfoA(tmpfile, SE_FILE_OBJECT,
#                                     OWNER_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION,
#                                     (PSID *)&owner, NULL, &pDacl, NULL, &pSD);
# -    todo_wine
#      ok(error == ERROR_SUCCESS, "GetNamedSecurityInfo failed with error %ld\n", error);
# -    if (error == ERROR_SUCCESS)
# -    {
# -        bret = GetAclInformation(pDacl, &acl_size, sizeof(acl_size), AclSizeInformation);
# -        ok(bret, "GetAclInformation failed\n");
# -        todo_wine
# -        ok(acl_size.AceCount == 0, "GetAclInformation returned unexpected entry count (%ld != 0).\n",
# -                                   acl_size.AceCount);
# -        LocalFree(pSD);
# -    }
# +    bret = GetAclInformation(pDacl, &acl_size, sizeof(acl_size), AclSizeInformation);
# +    ok(bret, "GetAclInformation failed\n");
# +    todo_wine
# +    ok(acl_size.AceCount == 0, "GetAclInformation returned unexpected entry count (%ld != 0).\n",
# +                               acl_size.AceCount);
# +    LocalFree(pSD);
#      CloseHandle(hTemp);
 
#  done:
diff --git a/server/fd.c b/server/fd.c
index 6cf8fb3ad06..d789895628b 100644
--- a/server/fd.c
+++ b/server/fd.c
@@ -1914,6 +1914,7 @@ struct fd *open_fd( struct fd *root, const char *name, struct unicode_str nt_nam
     int root_fd = -1;
     int rw_mode;
     char *path;
+    int do_chmod = 0;
     int created = (flags & O_CREAT);
 
     if (((options & FILE_DELETE_ON_CLOSE) && !(access & DELETE)) ||
@@ -1973,6 +1974,23 @@ struct fd *open_fd( struct fd *root, const char *name, struct unicode_str nt_nam
             if ((access & FILE_UNIX_WRITE_ACCESS) || (flags & O_CREAT))
                 fd->unix_fd = open( name, O_RDONLY | (flags & ~(O_TRUNC | O_CREAT | O_EXCL)), *mode );
         }
+        else if (errno == EACCES)
+        {
+            /* try to change permissions temporarily to open a file descriptor */
+            if (!(access & (FILE_UNIX_WRITE_ACCESS | FILE_UNIX_READ_ACCESS | DELETE)) &&
+                !stat( name, &st ) && st.st_uid == getuid() &&
+                !chmod( name, st.st_mode | S_IRUSR ))
+            {
+                fd->unix_fd = open( name, O_RDONLY | (flags & ~(O_TRUNC | O_CREAT | O_EXCL)), *mode );
+                *mode = st.st_mode;
+                do_chmod = 1;
+            }
+            else
+            {
+                set_error( STATUS_ACCESS_DENIED );
+                goto error;
+            }
+        }
 
         if (fd->unix_fd == -1)
         {
@@ -1981,6 +1999,8 @@ struct fd *open_fd( struct fd *root, const char *name, struct unicode_str nt_nam
                 set_error( STATUS_OBJECT_NAME_INVALID );
             else
                 file_set_error();
+
+            if (do_chmod) chmod( name, *mode );
             goto error;
         }
     }
@@ -1996,6 +2016,7 @@ struct fd *open_fd( struct fd *root, const char *name, struct unicode_str nt_nam
     closed_fd->unix_fd = fd->unix_fd;
     closed_fd->disp_flags = 0;
     closed_fd->unix_name = fd->unix_name;
+    if (do_chmod) chmod( name, *mode );
     fstat( fd->unix_fd, &st );
     *mode = st.st_mode;
 
-- 
2.45.2

